src/main/java/tfb/status/service/ShareResultsUploader.java

```diff
@@ -0,0 +1,305 @@
+   * @see #upload(ReadableByteChannel)
+   */
+  public ShareResultsJsonView upload(String resultsJson)
+          throws IOException, ShareResultsUploadException {
+    Objects.requireNonNull(resultsJson);
+
+    try (InputStream is = new ByteArrayInputStream(resultsJson.getBytes(UTF_8));
+         ReadableByteChannel in = Channels.newChannel(is)) {
+      return upload(in);
+    }
+  }
+
+  /**
+   * Upload the given byte channel containing the raw results JSON. Creates
+   * a temporary file from the bytes and passes it to {@link #upload(Path)}.
+   *
+   * @param in The byte channel containing the raw results JSON.
+   * @see #upload(Path)
+   */
+  public ShareResultsJsonView upload(ReadableByteChannel in)
+      throws IOException, ShareResultsUploadException {
+    Objects.requireNonNull(in);
+
+    // Copy the input to a temporary file.
+    Path tempFile = Files.createTempFile(/* prefix= */ "TFB_Share_Upload",
+        /* suffix= */ ".json");
+
+    try (WritableByteChannel out =
+             Files.newByteChannel(tempFile, WRITE, APPEND)) {
+      ByteStreams.copy(in, out);
+    }
+
+    return upload(tempFile);
+  }
+
+  /**
+   * Upload the given file containing raw results JSON. This first validates the
+   * size requirements: that the given file isn't too large, and that the share
+   * directory is not full. This method then validates the contents of the file,
+   * ensuring that it de-serializes to a {@link Results} object successfully,
+   * and that it contains a non-empty {@link Results#testMetadata}. If
+   * validation fails a {@link ShareResultsUploadException} is thrown. Otherwise
+   * this will create a new zip file in the {@link FileStore#shareDirectory()}
+   * containing the JSON file and return information about it.
+   *
+   * @param tempFile The file containing raw results JSON. After this method returns,
+   *                 this file is guaranteed to be deleted.
+   * @return A view containing information about the new file if validation passes
+   * and the new zip file was successfully created.
+   * @throws IOException If any network errors occur.
+   * @throws ShareResultsUploadException If validation of the JSON fails.
+   */
+  public ShareResultsJsonView upload(Path tempFile)
+      throws IOException, ShareResultsUploadException {
+    Objects.requireNonNull(tempFile);
+
+    try {
+      validateUploadSize(tempFile.toFile().length());
+      String validationError = validateNewFile(tempFile);
+      if (validationError != null) {
+        throw new ShareResultsUploadException(validationError);
+      }
+
+      String shareId = UUID.randomUUID().toString();
+      String fileName = shareId + ".json";
+      String zipFileName = shareId + ".zip";
+
+      Path permanentFile = fileStore.shareDirectory().resolve(zipFileName);
+      MoreFiles.createParentDirectories(permanentFile);
+
+      try (FileOutputStream fos = new FileOutputStream(permanentFile.toFile());
```

Comment:
Avoid calling `Path.toFile()`, which uses `java.io.File`.  I suggest avoiding the `ZipOutputStream` class completely and use the `java.nio.file` APIs instead.

```java
try (FileSystem zipFs = Files.newFileSystem(permanentFile)) {
  Path entry = zipFs.getPath(fileName);
  try (InputStream in = Files.newInputStream(tempFile);
       OutputStream out = Files.newOutputStream(entry, CREATE_NEW)) {
    in.transferTo(out);
  }
}
```

Reply:
Here's a slight correction to that example code for creating the zip file.

```java
try (FileSystem zipFs =
         FileSystems.newFileSystem(
             permanentFile,
             Map.of("create", "true")) {
  ...
}
```

[`jdk.zipfs` module documentation](https://docs.oracle.com/en/java/javase/13/docs/api/jdk.zipfs/module-summary.html)